---
title: "Class 14 (Lecture 15): Transcriptomics and the Analysis of RNA-seq data"
author: "Sarra Larif"
date: "2/20/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Install BiocManager
`install.packages("BiocManager")`
`BiocManager::install()`
For this class, you'll also need DESeq2:
`BiocManager::install("DESeq2")`
`library(DESeq2)`

##Import Today's Data

The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. The anti-inflammatory effects on airway smooth muscle (ASM) cells has been known for some time but the underlying molecular mechanisms are unclear.

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```
The rows in "counts" correspond to the columns in "metadata"
The numbers are the counts of reads (number of fragments mapped to the gene)

Number of genes we are working with:
```{r}
nrow(counts)
```

Average counts per gene for control samples and drug-treated samples then compare these values to see if there is a difference in the control and treated samples for a given gene:
```{r}
metadata$dex == "control"

#Want the ID number for where it says TRUE
metadata$id[metadata$dex == "control"]
#OR metadata[metadata$dex == "control", ]$id
```

This is our metadata info that tells us where our control and treated samples are (which columns in `counts`)
Access control only columns in the metadata
```{r}
control <- metadata[metadata$dex =="control",]
```
Now use the `$id` of the control data to access the control columns in `counts`
Do not want to just divide by 4 because if the data set size changes then it's wrong
```{r}
control.mean <- rowSums(counts[, control$id])/nrow(control)
#Add gene names back on to vector for clarity
names(control.mean) <- counts$ensgene
head(control.mean)
```

Now do the same thing for the `treated` columns 
```{r}
treated <- metadata[metadata$dex == "treated", ]
treated.mean <- rowSums(counts[, treated$id])/nrow(treated)
names(treated.mean) <- counts$ensgene
head(treated.mean)
```

Now want to make a dataframe with both mean sets
```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

Create a scatter plot to show the mean of the treated samples against the mean of control samples
```{r}
plot(meancounts, xlab = "Control", ylab = "Treated")
```
Use logs of data to spread out data to open that cluster 
```{r}
plot.default(meancounts, log = "xy", xlab = "log Control", ylab = "log Treated")
```
If above diagonal, it's upregulated in treated and if below diagonal the gene is downregulated when treated 

Now we want to divide the means so we can easily see upregulation or downregulation and make that a new column
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
NaN means it's trying to take the log of zero (the gene is not active in either treatment)

Remove any gene that has zero average counts.
```{r}
#arr.ind gives row and column indeces to tell you where in the table the value is zero
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

How many genes do I have left?
```{r}
nrow(mycounts)
```

Want to narrow down to the genes that are affected by the drug (if it's more than 2 (upregulated) or less than -2 (downregulated) we consider it a big change)
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
#These give logicals (TRUE/FALSE)
```

Want number of genes
```{r}
sum(up.ind)
sum(down.ind)
```

##Do this properly with DESeq2
```{r}
library(DESeq2)
```

Setup input for DESeq by specifying countdata, metadata and the experiment design (i.e. which comparisons we are interested in)
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

Run main function and put results back into the name we gave 
```{r}
dds <- DESeq(dds)
res <- results(dds)
```

###Volcano plot:
All the points are the genes and the 0 means no change and we are graphing the log of the fold change vs. the -log of the pvalue
To the right of 0 is upregulated genes and to the left of 0 is the downregulated genes 
```{r}
plot(res$log2FoldChange, res$padj)
#This gives an unusual plot but we are only interested in very low p values so we have to narrow
```

```{r}
res$sig <- res$padj<0.05 & abs(res$log2FoldChange)>2

# How many of each?
table(res$sig)
sum(is.na(res$sig))
```

```{r}
palette(c("gray", "blue"))
plot(res$log2FoldChange, -log(res$padj), col = res$sig+1, ylab = "-Log(P-value)", xlab="Log2(FoldChange)")
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.1), col="darkgray", lty=2)
#need -log to have data on bottom of page, not top
#We want to highlight 2-fold changes (log2>2) and statistical significance (p<0.01)
```

Reset the color palette
We could also setup a custom color vector indicating transcripts with large fold change and significant differences between conditions:
```{r}
palette("default")  


# Setup our custom point color vector. Everything is gray unless it is more than 2 or less than -2
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

#Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

Can do same with `ggplot`
```{r}
library(ggplot2)
ggplot(as.data.frame(res), aes(log2FoldChange, -log10(pvalue), col=sig)) + 
    geom_point() + 
    ggtitle("Volcano plot")
```

