---
title: "Class16: Pathway Analysis from RNA-Seq"
author: "Sarra Larif"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Lecture Notes

###Pathway Analysis
Used for experiment output of long list of genes; extract biological meaning (i.e. geneset enrichment)
Output as differentially expressed genes (DEG)
-genes from experiments, if there is some overlap with existing pathways then they are enriched pathways otherwise they shouldn't be pursued
Size bias because just by chance it's more likely to hit a big pathway than a small one

Same genes are stored with different tags in different databases so you have to translate between different languages 
Excel is manual and doesn't always work (especially with big files)

###GO
IEA: inferred from electric annotation (basically same as find a gene project)
As you go down graph, layers ar emore specific in describing the gene
More than half genes are inferred from find a gene approach instead of experiments 

##Hands-On Section
DESeq approach requires count data and colData

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:
Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

Here we read the two input files
```{r}
library(DESeq2)
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

colData <-  read.csv(metaFile, row.names = 1)
head(colData)

countData <-  read.csv(countFile, row.names = 1)
```

There are `r nrow(countData)` genes in this dataset and `r ncol(countData)` experiments (samples)
Adding the 'r' before the command makes it so the code runs when you knit the file

There is an extra column in `countData` for "length" that is not an experiment that we need to clean up so it does not mess with DESeq results 
```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
#To get everything except the first column you can do 2:7 or -1
```

We also want to remove data with 0 counts because there's no point in including that and diluting the data set

Take sum of rows and exclude rows that are 0
```{r}
countData <-  countData[rowSums(countData) != 0, ]
```

```{r}
nrow(countData)
```

###Running DESeq2

Now we run DESeq on our data

```{r}
dds <-  DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds <-  DESeq(dds)
```

Let's see our results
```{r}
res <-  results(dds, contrast=c("condition", "hoxa1_kd", "control_sirna"))
res
#Wanr p-value below threshold to be significant 
```

###Volcano Plot
Summary plot of results (plot of -2log of fold change vs. p-value)
```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (abs(res$log2FoldChange) < -2) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col = mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
#Everything above red line is significant 
#Everything outside of red lines has a significant fold change
#We want to extract our genes in those segments 
```

###Adding gene annotation:
Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.

I would install these with
`BiocManager::install("AnnotationDbi")` and
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

# What is availabvle in the org.Hs.eg.db dataset?
columns(org.Hs.eg.db)
```
Lets add extra annotation to our result object
```{r}
res$symbol <-  mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

head(res)
```

###Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we havea list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

First install packages:
`BiocManager::install( c("pathview", "gage", "gageData") )`

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data("sigmet.idx.hs")

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Make an input vector of our numeric quantity of interest i.e. fold change values and make sure it has ENTREZ gene names associated with it
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now run **gage** pathway analysis with this input vector and our KEGG genesets
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

What is in our results? Use `attributes()` to see what is returned
```{r}
attributes(keggres)

# Look at the first few down (less) pathways
head(keggres$less)

#hsa number is the accession ID for KEGG
```

Now we use KEGG pathway ID from our top his along with out `foldchange` vector to view our results as a more useful figure
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
#This outputs figures to files tab that creates an image when clicked
#Many DEGs overlap with this pathway
```

Include graphics in knitted markdown
```{r}
knitr::include_graphics("hsa04110.pathview.png")
```



